### DISCUSSION TITLE
Images Best Practices

### DISCUSSION THREAD

**Discussion by malfonso0** [fastapi/fastapi#6706] [Questions] [answered]
hi, i am developing a object detection projects on jetsons Xavier NX. where i need to monitor multiple cameras, and do some object detection on each frame of each camera
 :( Sorry for the long question, im trying to introduce thr problem/objective

At first... i started with only one camera... so all my code went on the same script.. something like

model = loadModel()
while True
   frame = getFrame()
   detections = model.detect(frame)
   ...

For this simple script.. i had almost 70FPS (~15 ms) ( model is a tinyyolo4-416.. )
Then i started with the multicamera approach... i tryed threading/multiprocess.. and worked, but i think that lacks separation of concerns.. and my code was a bit messy.

in order to do separation of concerns.. i haave created two scripts/dockers.
1. An Api using fastapi. which loads the model, and has the predict endpoint for prediction
2. CameraProcessor.. which, as the name goes, its a python script that takes images from a camera, and it then call the API for detectin things on the frame.

when i started testing this, again with only one camera.. my FPS went down from 70 FPS to 30FPS (~ 30ms)

Here my questions
* **What is the recomended/best practices to send images to the API?** 
right now.. im encoding the numpy image( cv2.imencode then to base64) and sending and decoding on the other side... i have done some profiling.. and the encoding/decoding takes approx 4 ms each.. wich would reduce my FPS to ~50

* **Is there any "configuration" that i can do ? for reducing this gap ? or this is the expected latency?**
thanks

**Comment by baby5** [comment]
in your case, not use API

**Comment by malfonso0** [comment]
thanks @baby5  for your response
could you please.. explain why?..

**Comment by STeveShary** [comment, marked as answer]
@malfonso0 It sounds like latency is important to you.  One thought is that HTTP calls through FastAPI may not meet your needs.  Before you jump to far into an approach of using FastAPI, I would recommend that you understand the amount of time it will take to send an image over HTTP.  Below is a very simple FastAPI app that will return the name of the file sent:

```from fastapi import FastAPI, File, UploadFile

app = FastAPI()

@app.post("/image")
async def image(image: UploadFile = File(...)):
    return {"filename": image.filename}

```

I expect that you will find that the time to send an image to FastAPI will be > 50ms.   It is easier to scale the number of images that can be processed but you will find that the latency may be far too much for you.

Best of luck!

**Comment by malfonso0** [comment]
thanks @STeveShary  for your words.
 seems you are correct, it doesnt matter what i do... i always have 20-50ms latency. so this is "what's excepted" and there is not much room for improve
thanks

**Comment by Talgin** [comment]
If API style is what you are loking for you can use Triton Inference Server with grpc it will handle a lot of concurrent requests. IMHO, better maybe be to process images in "real time", using TensorRT or smth like it and send metadata through Kafka or other message queue and display results reading from it.
